### PR TITLE
Add separate skill table

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/Command/SkillsDataTransferCommand.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+{
+    /**
+     * @var \Doctrine\ORM\EntityManager
+     */
+    private $manager;
+
+    protected function configure()
+    {
+        $this
+            ->setName('app:skills:data_transfer')
```
Comment (hhamon) [inline src/Command/SkillsDataTransferCommand.php]: Should be `app:skills:data-transfer`

File: src/Command/SkillsDataTransferCommand.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            ->setDescription('Transfer the skill data to new skill table.')
```
Comment (hhamon) [inline src/Command/SkillsDataTransferCommand.php]: `'Transfer the skill data to the new skill table.'`

File: src/Entity/Skill.php [hunk 3]
```diff
@@ -25,27 +31,26 @@ class Skill
     /**
      * @var string
      *
-     * @ORM\Column
+     * @ORM\Column(unique=true)
```
Comment (hhamon) [inline src/Entity/Skill.php]: There is already a `@ORM\UniqueConstraint` annotation at the class level.

File: app/migrations/Version20170725171157.php [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+namespace Migrations;
+
+use Doctrine\DBAL\Migrations\AbstractMigration;
+use Doctrine\DBAL\Schema\Schema;
+
+class Version20170725171157 extends AbstractMigration
+{
+    public function up(Schema $schema)
+    {
+        $this->addSql('DROP TABLE member_summary_skills');
```
Comment (HeahDude) [inline app/migrations/Version20170725171157.php]: Why another migration for this?

File: app/migrations/Version20170725171157.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    }
+
+    public function down(Schema $schema)
+    {
+        $this->addSql('DROP TABLE summary_skills');
```
Comment (HeahDude) [inline app/migrations/Version20170725171157.php]: This looks wrong, and it should be part of the other migration.

File: src/Command/SkillsDataTransferCommand.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    protected function initialize(InputInterface $input, OutputInterface $output)
+    {
+        $this->manager = $this->getContainer()->get('doctrine.orm.entity_manager');
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $sqlFromCommitteeMembership = <<<'SQL'
+          ALTER TABLE skills ADD slug VARCHAR(255) NOT NULL;
```
Comment (HeahDude) [inline src/Command/SkillsDataTransferCommand.php]: Why isn't this part of the migration file?

File: src/Command/SkillsDataTransferCommand.php [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+          DROP FUNCTION ;
+          
+          ALTER TABLE skills DROP slug;
+          
+          
+          
+          
+          
+          
+          INSERT INTO projection_referent_managed_users
```
Comment (HeahDude) [inline src/Command/SkillsDataTransferCommand.php]: All this query looks really weird, I guess this is part of the WIP.

File: src/DataFixtures/ORM/LoadSummaryData.php [hunk 8]
```diff
@@ -115,20 +116,13 @@ public function load(ObjectManager $manager)
         $summary1->addLanguage($language13);
 
         // Skills
-        $skill11 = new Skill();
-        $skill11->setName('Software');
+        $skill11 = $manager->getRepository(Skill::class)->findOneBy(['name' => LoadSkillData::SKILLS['S001']]);
+        $skill12 = $manager->getRepository(Skill::class)->findOneBy(['name' => LoadSkillData::SKILLS['S002']]);
+        $skill13 = $manager->getRepository(Skill::class)->findOneBy(['name' => LoadSkillData::SKILLS['S003']]);
+        $skill14 = $manager->getRepository(Skill::class)->findOneBy(['name' => LoadSkillData::SKILLS['S004']]);
```
Comment (HeahDude) [inline src/DataFixtures/ORM/LoadSummaryData.php]:
```php
foreach ($manager->getRepository(Skill::class)->findBy(['name' => [
    LoadSkillData::SKILLS['S001']],
    LoadSkillData::SKILLS['S002']],
    LoadSkillData::SKILLS['S003']],
    LoadSkillData::SKILLS['S004']],
]) as $skill) {
    $summary1->addSkill($skill);
}
```

File: src/DataFixtures/ORM/LoadSummaryData.php [hunk 9]
```diff
@@ -184,25 +178,14 @@ public function load(ObjectManager $manager)
...
-
-        $skill24 = new Skill();
-        $skill24->setName('Outils médias');
-        $skill24->setSummary($summary2);
-        $manager->persist($skill24);
+        $skill21 = $manager->getRepository(Skill::class)->findOneBy(['name' => LoadSkillData::SKILLS['S005']]);
+        $skill22 = $manager->getRepository(Skill::class)->findOneBy(['name' => LoadSkillData::SKILLS['S006']]);
+        $skill23 = $manager->getRepository(Skill::class)->findOneBy(['name' => LoadSkillData::SKILLS['S007']]);
+        $skill24 = $manager->getRepository(Skill::class)->findOneBy(['name' => LoadSkillData::SKILLS['S008']]);
+        $summary2->addSkill($skill21);
```
Comment (HeahDude) [inline src/DataFixtures/ORM/LoadSummaryData.php]: Same here.

File: src/Form/SummaryType.php [hunk 10]
```diff
@@ -50,10 +52,16 @@ class SummaryType extends AbstractType
      */
     private $glide;
 
-    public function __construct(Filesystem $storage, Server $glide)
+    /**
+     * @var EntityManagerInterface
```
Comment (HeahDude) [inline src/Form/SummaryType.php]: No need.

File: src/Entity/Skill.php [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+    /**
+     * @var string
+     *
+     * @ORM\Column(unique=true)
+     *
+     * @Assert\NotBlank
+     * @Assert\Length(min=2, max=200)
+     */
+    private $name = '';
```
Comment (HeahDude) [inline src/Entity/Skill.php]: I don't see any slug property anymore, the unique constraint is not enough if the value is not normalised at some point, maybe using `mb_strtolower`?

Also the initialization is pointless due to the constructor.

File: src/Entity/Skill.php [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+     */
+    private $id;
+
+    /**
+     * @var string
+     *
+     * @ORM\Column(unique=true)
+     *
+     * @Assert\NotBlank
+     * @Assert\Length(min=2, max=200)
```
Comment (HeahDude) [inline src/Entity/Skill.php]: No length of 200 for the column?

File: src/Entity/Skill.php [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    /**
+     * @ORM\ManyToMany(targetEntity="AppBundle\Entity\Summary", mappedBy="skills")
+     *
+     * @var Summary[]
+     */
+    private $summaries;
+
+    public function __construct(?string $name = null)
+    {
+        if ($name) {
```
Comment (HeahDude) [inline src/Entity/Skill.php]: The `if` could be removed in favor of casting the value to string.

File: src/Form/SummaryType.php [hunk 14]
```diff
@@ -138,6 +146,22 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                        $summary = $event->getForm()->getData();
+                        $skills = $event->getData()->getSkills();
+
+                        foreach ($skills as $skill) {
+                            if ($skill->getId()) {
+                                continue;
+                            }
+
+                            if ($existingSkill = $this->entityManager->getRepository(Skill::class)->findOneBy(['name' => $skill->getName()])) {
+                                $summary->replaceSkill($skill, $existingSkill);
```
Comment (HeahDude) [inline src/Form/SummaryType.php]: I'm not sure this method is relevant if the name is not normalized first.

File: src/Repository/SkillRepository.php [hunk 15]
```diff
@@ -2,35 +2,20 @@
...
-            ->leftJoin('s.summary', 'summary')
-            ->where('s.slug LIKE :slug')
-            ->andWhere('summary.member != :member')
-            ->andWhere($qb->expr()->notIn('s.slug', $qbUserSkills->getDQL()))
+            ->where('s.name LIKE :name')
             ->setParameters([
-                'slug' => $term.'%',
-                'member' => $user,
-                'user' => $user,
+                'name' => $term.'%',
```
Comment (HeahDude) [inline src/Repository/SkillRepository.php]: The term should be normalized.

File: src/Command/SkillsDataTransferCommand.php (hunk 7 above)
Comment (illusionOfParadise) [inline src/Command/SkillsDataTransferCommand.php]: Yes

**Pull request merged**